tests/avifmetadatatest.c

```diff
@@ -0,0 +1,221 @@
+        printf("ERROR: avifDecoderReadMemory() failed\n");
+        goto cleanup;
+    }
+    success = AVIF_TRUE;
+cleanup:
+    if (decoder != NULL) {
+        avifDecoderDestroy(decoder);
+    }
+    return success;
+}
+
+//------------------------------------------------------------------------------
+
+// Returns true if the decoded output metadata matches the input metadata.
+static avifBool metadataIsEqual(const avifRWData * inputItem, const avifRWData * outputItem)
+{
+    return (outputItem->size == inputItem->size) && (memcmp(outputItem->data, inputItem->data, inputItem->size) == 0);
+}
+
+// Encodes, decodes then compares the metadata of the input and decoded images.
+static avifBool encodeDecode(const avifImage * image)
+{
+    avifBool success = AVIF_FALSE;
+    avifRWData encodedAvif = AVIF_DATA_EMPTY;
+    avifImage * decodedImage = NULL;
+    if (!encode(image, &encodedAvif)) {
+        goto cleanup;
+    }
+    if (!decode(&encodedAvif, &decodedImage)) {
+        goto cleanup;
+    }
+
+    if (!metadataIsEqual(&image->icc, &decodedImage->icc)) {
+        printf("ERROR: icc mismatch\n");
+        goto cleanup;
+    }
+    if (!metadataIsEqual(&image->exif, &decodedImage->exif)) {
+        printf("ERROR: Exif metadata mismatch\n");
+        goto cleanup;
+    }
+    if (!metadataIsEqual(&image->xmp, &decodedImage->xmp)) {
+        printf("ERROR: XMP metadata mismatch\n");
+        goto cleanup;
+    }
+    success = AVIF_TRUE;
+cleanup:
+    avifRWDataFree(&encodedAvif);
+    if (decodedImage) {
+        avifImageDestroy(decodedImage);
+    }
+    return success;
+}
+
+//------------------------------------------------------------------------------
+
+// Encodes, decodes then verifies that the output metadata matches the input
+// metadata defined by the arguments.
+static avifBool encodeDecodeMetadataItems(avifBool useICC, avifBool useExif, avifBool useXMP)
+{
+    avifBool success = AVIF_FALSE;
+    avifImage * image;
+    if (!createImage1x1(&image)) {
+        goto cleanup;
+    }
+
+    if (useICC) {
+        avifImageSetProfileICC(image, sampleICC, sampleICCSize);
+    }
+    if (useExif) {
+        avifImageSetMetadataExif(image, sampleExif, sampleExifSize);
+    }
+    if (useXMP) {
+        avifImageSetMetadataXMP(image, sampleXMP, sampleXMPSize);
+    }
+
+    if (!encodeDecode(image)) {
+        goto cleanup;
+    }
+    success = AVIF_TRUE;
+cleanup:
+    if (image) {
+        avifImageDestroy(image);
+    }
+    return success;
+}
+
+int main(void)
+{
+    if (!encodeDecodeMetadataItems(/*useICC=*/AVIF_TRUE, /*useExif=*/AVIF_FALSE, /*useXMP=*/AVIF_FALSE) ||
+        !encodeDecodeMetadataItems(/*useICC=*/AVIF_FALSE, /*useExif=*/AVIF_TRUE, /*useXMP=*/AVIF_FALSE) ||
+        !encodeDecodeMetadataItems(/*useICC=*/AVIF_FALSE, /*useExif=*/AVIF_FALSE, /*useXMP=*/AVIF_TRUE)) {
+        return EXIT_FAILURE;
+    }
+    // TODO: Negative test
+    // TODO: Multi xmp test
+    return EXIT_SUCCESS;
```

Comment:
Nit: Print a message to indicate the test passed. Otherwise one needs to check the exit status.